game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/change/RemoveUnprotectedUnits.java

```diff
@@ -0,0 +1,156 @@
+    }
+  }
+
+  private void checkUndefendedTransports(
+      final IDelegateBridge bridge, final BattleState.Side side) {
+    final GamePlayer player =
+        side == BattleState.Side.OFFENSE ? battleState.getAttacker() : battleState.getDefender();
+    // if we are the attacker, we can retreat instead of dying
+    if (attackerHasRetreat(side)) {
+      return;
+    }
+    final List<Unit> alliedTransports = getAlliedTransports(player);
+    if (alliedTransports.isEmpty()) {
+      return;
+    }
+    final Collection<Unit> alliedUnits = getAlliedUnits(player);
+    // If transports are unescorted, check opposing forces to see if the Trns die automatically
+    if (alliedTransports.size() == alliedUnits.size()) {
+      final Collection<Unit> enemyUnits = getEnemyUnitsThatCanFire(player);
+      // If there are attackers set their movement to 0 and kill the transports
+      if (!enemyUnits.isEmpty()) {
+        final Change change =
+            ChangeFactory.markNoMovementChange(
+                CollectionUtils.getMatches(enemyUnits, Matches.unitIsSea()));
+        bridge.addChange(change);
+        battleActions.remove(
+            alliedUnits, bridge, battleState.getBattleSite(), side == BattleState.Side.DEFENSE);
+      }
+    }
+  }
+
+  private boolean attackerHasRetreat(final BattleState.Side side) {
+    return side == BattleState.Side.OFFENSE
+        && (!battleState.getAttackerRetreatTerritories().isEmpty()
+            || battleState.getAttackingUnits().stream().anyMatch(Matches.unitIsAir()));
+  }
+
+  private List<Unit> getAlliedTransports(final GamePlayer player) {
+    final Predicate<Unit> matchAllied =
+        Matches.unitIsTransport()
+            .and(Matches.unitIsNotCombatTransport())
+            .and(Matches.isUnitAllied(player, battleState.getGameData()))
+            .and(Matches.unitIsSea());
+    return CollectionUtils.getMatches(battleState.getBattleSite().getUnits(), matchAllied);
+  }
+
+  private Collection<Unit> getAlliedUnits(final GamePlayer player) {
+    final Predicate<Unit> alliedUnitsMatch =
+        Matches.isUnitAllied(player, battleState.getGameData())
+            .and(Matches.unitIsNotLand())
+            .and(Matches.unitIsSubmerged().negate());
+    return CollectionUtils.getMatches(battleState.getBattleSite().getUnits(), alliedUnitsMatch);
+  }
+
+  private Collection<Unit> getEnemyUnitsThatCanFire(final GamePlayer player) {
+    final Predicate<Unit> enemyUnitsMatch =
+        Matches.unitIsNotLand()
+            .and(Matches.enemyUnit(player, battleState.getGameData()))
+            .and(Matches.unitIsSubmerged().negate())
+            .and(Matches.unitCanAttack(player));
+    return CollectionUtils.getMatches(battleState.getBattleSite().getUnits(), enemyUnitsMatch);
+  }
+
+  private void checkUnprotectedUnits(final IDelegateBridge bridge, final BattleState.Side side) {
+    // if we are the attacker, we can retreat instead of dying
+    if (attackerHasRetreat(side)) {
```

Comment:
What do you think about something that is more like:
```
if(!attackerCanRetreat(side)) {
   removeUnprotectedUnits(side);
}

```

Having control flow at the outer layers is generally a good thing. A series of void method calls is not a great abstraction, it's cognitively equivalent to adding comments and inlining the methods. A reader being able to skip sections of code because they are tracing control flow is on the other hand very valuable beyond demonstrating the flow of control at the outset. That pattern is called "[imperative shell, functional core](https://www.destroyallsoftware.com/screencasts/catalog/functional-core-imperative-shell)". 

I also suggest that as it took quite a bit of 'groking' to determine that this method is largely just a lot of conditionals then we do the remove operation. I think there is also a mixing of layers of abstraction code smell here too, eg:
```
- high level boolean check
- low level boolean check
- low level boolean check
- high level fetch of unit collection
- low level removal of units
```

With the above you keep the high level boolean check and the control flow is extracted, and you can also potentially inline `getUnprotectedUnits` into `removeUnprotectedUnits` which I think then makes a lot of sense and keeps the abstractions at a consistent level (and there are just as many methods, but each one is more focused in what they do).

Reply:
I had basically copied the original code and extracted out larger pieces so that it was more readable but you are right, it is still hard to read.  So, I've restructured the code to follow the [return early pattern](https://medium.com/swlh/return-early-pattern-3d18a41bba8).